sys/kern/lockdep.c

```diff
@@ -32,7 +32,12 @@ typedef struct lock_class_link {
 static spin_t main_lock = SPIN_INITIALIZER(0);
 
 #define CLASSHASH_SIZE 64
-#define CLASSHASH(key) ((uintptr_t)(key) % CLASSHASH_SIZE)
+/* We have to divide the key by the alignment of lock_class_key_t to prevent the
+   periodicity.
+   A key which is a multiple of 4 results in a hash which is also a multiple of
+   that number. So in the end we would only use every 4th bucket. */
+#define CLASSHASH(key)                                                         \
+  (((uintptr_t)(key) / alignof(lock_class_key_t)) % CLASSHASH_SIZE)
```

Comment:
I would also add a static assertion that `alignof(lock_class_mapping_t) >= alignof(lock_class_key_t)`, since `key` can also be a pointer to `lock_class_mapping_t`.

Reply:
...but `lock_class_mapping_t` is aligned to 8 bytes on aarch64, so lock classes of statically allocated locks will have a `key` that's a multiple of 8, not 4... we could fix this by requiring `lock_class_key_t` to have the same alignment as `lock_class_mapping_t`, but that will waste more space for keys...

Reply:
I wouldn't care much about that extra memory. I've changed the type of `lock_class_key_t` to `uintptr_t` so the alignment will be always the length of the word.

Reply:
Ok, good enough.